### DISCUSSION TITLE
No ARM Docker image available

### DISCUSSION THREAD

**Discussion by vaisest** [RogueOneEcho/caesura#97] [Ideas]
```
 ! caesura The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested 0.0s
Attaching to caesura
caesura  | exec /bin/caesura: exec format error
caesura exited with code 1
```

I haven't tried multiarch builds myself, so I hope I'm not incorrect, but I believe it's not very hard to implement: https://www.docker.com/blog/how-to-rapidly-build-multi-architecture-images-with-buildx/

It'd be nice if there was an ARM image. Building the Rust binary takes quite a long time on a weak CPU: 
```
 => [builder  7/10] RUN cargo build --release --locked                                956.3s
 => [builder  8/10] COPY . /app                                                       0.4s
 => [builder  9/10] RUN cargo set-version 0.0.0                                       2.3s
 => [builder 10/10] RUN cargo build --release                                         335.8s
```

**Comment by RogueOneEcho** [comment]
tl;dr
I'll look in to it but it's not a priority atm, if there's clear demand I'll revaluate.

---

I can appreciate the frustration, I recall giving up on ARM servers for this exact reason.

Are your requesting this to run on a Raspberry PI or is it a low cost dedicated server?

From the link it does look like it might be as simple as 
- [ ] adding `--platform linux/amd64,linux/arm64` to the buildx command. 

However, the CI build times are already frustratingly long -  despite my attempts to cache as much as possible -  and I suspect a multi architecture build will double the length so I need to weigh up the extended build times against the use ARM builds will actually get.

**Reply by vaisest** [reply]
> Are your requesting this to run on a Raspberry PI

This is on a Raspberry Pi 4 indeed. I use it as a NAS

> I suspect a multi architecture build will double the length

They seem to be built in parallel, but since the arm64 build is non-native (using QEMU), it's certainly slow and it becomes a bottleneck. A test build took more than 20 minutes, so I definitely understand your point lol

**Reply by RogueOneEcho** [reply]
Even if they're technically built in parallel they share the resources of a single GitHub Actions runner so there's no real benefit from buildx parallelism. 

https://github.com/RogueOneEcho/caesura/actions/runs/12105463159/job/33750108326

`docker-build` increased from 4m 18s to 36m 31s. That's insane.

Option 2:

It may be possible to:
1. build amd64 and arm64 as separate GitHub Actions jobs 
2. publish amd64 as soon as it completes
3. after arm64 completes to merge the amd64 and arm64 image manifests and republish the image.

That way arm64 gets dedicated resources for true parallelism and the build time doesn't delay amd64 being published.

Option 3:

Or skip step 3 and just publish the arm64 image with a different tag. That's a lot cleaner than merging manifests but it does ignore Docker's multi arch capabilities.

For example:
```bash
docker run ghcr.io/rogueoneecho/caesura:arm64 --help
```

Option 4:
Build an arm64 binary as a cargo/rustup target and just copy that to a docker image.
It reduces transparency because the build process is partially obscured but I suspect it would be a lot quicker.

**Reply by RogueOneEcho** [reply]
GitHub Actions will have native ARM  runners sometime in Q1 2025 which should simplify this:

https://github.com/github/roadmap/issues/970
https://github.com/actions/runner-images/issues/10820

**Reply by RogueOneEcho** [reply]
Completed build of ARM binaries in ca41eb1e
Completed build of multi-arch docker image in 473cc3f1

Comment above noted that using buildx it took 36m 31s. Using the native GitHub runners that are now available intel as 5m in arm 6m. These run in parallel so the full CI/CD process isn't really any longer.

**Comment by RogueOneEcho** [comment]
Completed build of ARM binaries in https://github.com/RogueOneEcho/caesura/commit/ca41eb1e3603b32d8132dd9df86636fa5cf86dc4
Completed build of multi-arch docker image in https://github.com/RogueOneEcho/caesura/commit/473cc3f1bd1a8dbe569d2bda4286c195294dd2a5
